File: app/src/main/java/de/markusfisch/android/shadereditor/hardware/AbstractListener.java
package de.markusfisch.android.shadereditor.hardware;

import de.markusfisch.android.shadereditor.app.ShaderEditorApp;

import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;

public abstract class AbstractListener implements SensorEventListener {
	long last = 0;

	private final SensorManager sensorManager;

	private boolean listening = false;
	private Sensor sensor;

	AbstractListener(Context context) {
		sensorManager = (SensorManager) context.getSystemService(
				Context.SENSOR_SERVICE);
	}

	public void unregister() {
		if (sensor == null || !listening) {
			return;
		}

		sensorManager.unregisterListener(this);
		listening = false;
	}

	@Override
	public void onAccuracyChanged(Sensor sensor, int accuracy) {
	}

	@Override
	public void onSensorChanged(SensorEvent event) {
		last = event.timestamp;
	}

	boolean register(int type) {
		if (listening || sensorManager == null || (sensor == null &&
				(sensor = sensorManager.getDefaultSensor(type)) == null)) {
			return false;
		}

		last = 0;
		listening = sensorManager.registerListener(this, sensor,
				ShaderEditorApp.preferences.getSensorDelay());

		return listening;
	}
}

File: app/src/main/java/de/markusfisch/android/shadereditor/hardware/AbstractListener.kt
package de.markusfisch.android.shadereditor.hardware

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import de.markusfisch.android.shadereditor.app.ShaderEditorApp

abstract class AbstractListener(context: Context) : SensorEventListener {

    var last: Long = 0
        protected set

    private val sensorManager: SensorManager? =
        context.getSystemService(Context.SENSOR_SERVICE) as SensorManager?

    private var listening = false

    private var sensor: Sensor? = null

    fun unregister() {
        if (sensor != null && listening) {
            sensorManager?.unregisterListener(this)
            listening = false
        }
    }

    override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {
        // No implementation required
    }

    override fun onSensorChanged(event: SensorEvent) {
        last = event.timestamp
    }

    fun register(type: Int): Boolean {
        if (listening) return false
        sensor = sensor ?: sensorManager?.getDefaultSensor(type) ?: return false

        last = 0
        listening = sensorManager?.registerListener(
            this, sensor, ShaderEditorApp.preferences.sensorDelay
        ) ?: false

        return listening
    }
}